parlai/agents/transformer/image_polyencoder.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# hack to make sure -m transformer/generator works as expected
+"""
+Poly-encoder agent that ingests image features.
+"""
+
+from typing import Any, Dict, List
+
+import torch
+
+from parlai.agents.image_seq2seq.modules import ContextWithImageEncoder
+from parlai.agents.transformer.modules import get_n_positions_from_options
+from parlai.agents.transformer.polyencoder import PolyencoderAgent, PolyEncoderModule
+from parlai.core.message import Message
+from parlai.core.torch_agent import Batch
+
+DEFAULT_IMAGE_FEATURES_DIM = 2048
+
+
+class ImagePolyencoderAgent(PolyencoderAgent):
+    """
+    Poly-encoder Agent that ingests image features.
+
+    Agent that allows encoding image features and adding or concatenating them to the
+    context encoding.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add command-line arguments specifically for this agent.
+        """
+        super(ImagePolyencoderAgent, cls).add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Image Encoder Args')
+        agent.add_argument(
+            '--polyencoder-image-encoder-num-layers',
+            type=int,
+            default=1,
+            help='Number of linear layers to encode image features with in the context',
+        )
+        agent.add_argument(
+            '--polyencoder-image-features-dim',
+            type=int,
+            default=DEFAULT_IMAGE_FEATURES_DIM,
+            help='For passing in image features of the given dim in the context',
+        )
+        agent.add_argument(
+            '--polyencoder-image-combination-mode',
+            type=str,
+            default='prepend',
+            choices=['add', 'append', 'prepend'],
+            help='How to combine image embedding (if used) with context embedding',
+        )
+        agent.set_defaults(reduction_type=None)
+        # This agent doesn't support any encoder output reductions
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.image_features_dim = opt.get(
+            'polyencoder_image_features_dim', DEFAULT_IMAGE_FEATURES_DIM
+        )
+
+    def build_model(self, states=None):
+        """
+        Return built model.
+        """
+        return ImagePolyencoderModule(self.opt, self.dict, self.NULL_IDX)
+
+    def batchify(self, obs_batch: List[Message], sort: bool = False) -> Batch:
+        """
+        Override to handle image features.
+        """
+        batch = super().batchify(obs_batch, sort)
+
+        def _process_image_features(features: torch.Tensor) -> torch.Tensor:
+            assert features.size() == (self.image_features_dim,)
```

Comment:
have you tried this with other image features? e.g. `resnet152` or `resnext101_32x48d_wsl`? these return 4d tensors with size `[1, IMAGE_FEATURES_DIM, 1, 1]`, so you'll need to extract the tensor appropriately.

Reply:
I've fixed this in torch_image_agent.py